File: dp_educative/1_knapsack/01-knapsack_button_up_optimized.py
from pprint import pprint


def solve_knapsack(profits, weights, capacity):
    n = len(profits)
    if capacity <= 0 or n == 0 or len(weights) != n:
        return 0

    dp = [0 for i in range(capacity+1)]

    for c in range(0, capacity+1):
        if weights[0] <= c:
            dp[c] = profits[0]

    for i in range(1, n):
        for j in range(capacity, -1, -1):
            if weights[i] <= j:
                dp[j] = max(dp[j], profits[i] + dp[j-weights[i]])
    return dp[capacity]


print(solve_knapsack(tuple([1, 6, 10, 16]), tuple([1, 2, 3, 5]), 7))
print(solve_knapsack(tuple([1, 6, 10, 16]), tuple([1, 2, 3, 5]), 6))

File: dp_educative/1_knapsack/01-knapsack_button_up_optimized.cpp
#include "bits/stdc++.h"

using namespace std;

class Knapsack {

public:
  int solveKnapsack(const vector<int> &profits, const vector<int> &weights,
                    int capacity) {
      vector<int> dp = vector<int>(capacity + 1, 0);
    for (int c = 1; c < capacity + 1; ++c) {
      if (c >= weights[0]) {
        dp[c] = profits[0];
      }
    }

    for (int i = 1; i < capacity+1; ++i) {
      for (int j = capacity; j > -1; --j) {
        if(j>=weights[i]){
          dp[j] = max(dp[j], profits[i] + dp[j - weights[i]]);
        }
      }
    }
    return dp[capacity];
  }
};

int main(int argc, char *argv[]) {
  Knapsack ks;
  vector<int> profits = {1, 6, 10, 16};
  vector<int> weights = {1, 2, 3, 5};
  int maxProfit = ks.solveKnapsack(profits, weights, 7);
  cout << "Total knapsack profit ---> " << maxProfit << endl;
  maxProfit = ks.solveKnapsack(profits, weights, 6);
  cout << "Total knapsack profit ---> " << maxProfit << endl;
}